src/librustdoc/html/render/mod.rs

```diff
@@ -1037,22 +1037,27 @@ themePicker.onblur = handleThemeButtonsBlur;
         cx.shared.fs.write(&dst, v.as_bytes())?;
     }
 
-    // Update the search index
+    // Update the search index and crate list.
     let dst = cx.dst.join(&format!("search-index{}.js", cx.shared.resource_suffix));
     let (mut all_indexes, mut krates) = try_err!(collect_json(&dst, &krate.name.as_str()), &dst);
     all_indexes.push(search_index);
+    krates.push(krate.name.to_string());
 
     // Sort the indexes by crate so the file will be generated identically even
     // with rustdoc running in parallel.
     all_indexes.sort();
     {
         let mut v = String::from("var searchIndex = JSON.parse('{\\\n");
         v.push_str(&all_indexes.join(",\\\n"));
-        // "addSearchOptions" has to be called first so the crate filtering can be set before the
-        // search might start (if it's set into the URL for example).
-        v.push_str("\\\n}');\naddSearchOptions(searchIndex);initSearch(searchIndex);");
+        v.push_str("\\\n}');\ninitSearch(searchIndex);");
         cx.shared.fs.write(&dst, &v)?;
     }
+
+    let crate_list_dst = cx.dst.join(&format!("crates{}.js", cx.shared.resource_suffix));
+    let crate_list =
+        format!("window.ALL_CRATES = [{}];", krates.iter().map(|k| format!("\"{}\"", k)).join(","));
```

Comment:
I think this is still invalid: you'll overwrite any crate that isn't a dependency of the current one (when you are in a workspace for example, or simply documenting unrelated code in the same location). You need to extract the crate list and then update it like we do in the search index (we don't overwrite, we update, which is very different!).

Reply:
I may be misunderstanding, but I think this does the right thing. the `krates` list comes from `collect_json`, which reads it from the `search-index{}.js` as it currently stands. So as each run of `rustdoc` happens, new crates are added to `search-index{}.js`, and `crates{}.js` is updated based on the latest contents of `search-index{}.js`.

Reply:
As a test, I checked out https://github.com/servo/rust-url, which is a workspace containing 5 members:

```
[workspace]
members = ["url", "form_urlencoded", "idna", "percent_encoding", "data-url"]
```

Building docs with this branch produces this (truncated) search-index:

```
var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

And this crates.js:
```
window.ALL_CRATES = ["data_url","form_urlencoded","idna","matches","percent_encoding","tinyvec","tinyvec_macros","unicode_bidi","unicode_normalization","url"];
```

Building docs with nightly produces this (truncated) search-index:

```var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

Reply:
And what happens if you now document another non-related crate and make the output into this folder? The `search-index.js` will be updated and the `crates.js` will very likely only contain this new crate.

Reply:
This also works. I just tested by doing:

```
cd rust/ureq
cargo doc
cd ../conduit-cookie
cargo doc --target-dir ../ureq/target/
```

The resulting crates.js includes ureq, conduit-cookie, and all the dependencies of each:

```
window.ALL_CRATES = ["aead","aes","aes_gcm","aes_soft","base64","block_buffer","bytes","cfg_if","chunked_transfer","cipher","conduit","conduit_cookie","conduit_middleware","const_fn","cookie","cpuid_bool","crypto_mac","ctr","digest","fnv","form_urlencoded","generic_array","getrandom","ghash","hkdf","hmac","http","idna","itoa","libc","log","matches","once_cell","opaque_debug","percent_encoding","polyval","ppv_lite86","proc_macro2","proc_macro_hack","quote","rand","rand_chacha","rand_core","ring","rustls","sct","sha2","spin","standback","subtle","syn","time","time_macros","time_macros_impl","tinyvec","tinyvec_macros","typenum","unicode_bidi","unicode_normalization","unicode_xid","universal_hash","untrusted","ureq","url","webpki","webpki_roots"];
```

Reply:
I just checked, you are absolutely right!